### PR TITLE
Introduce collateral registry

### PR DESCRIPTION
**Pull request by bcngod**

### CODE REVIEW

File: contracts/CollateralRegistry.sol [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+
+        uint id = collateralId[asset];
+        require(id != 0 || _collaterals[0] == asset, "Unit Protocol: DOES_NOT_EXIST");
+
+        delete collateralId[asset];
+
+        uint lastId = _collaterals.length - 1;
+        address lastCollateral = _collaterals[lastId];
+        _collaterals[id] = lastCollateral;
+        collateralId[lastCollateral] = id;
```
Comment (Eenae) [inline contracts/CollateralRegistry.sol:52]: Incorrect if `lastId == id`.

File: contracts/ParametersBatchUpdater.sol [hunk 2]
```diff
@@ -7,24 +7,33 @@ pragma solidity ^0.7.1;
 
 import "./vault-managers/VaultManagerParameters.sol";
 import "./Vault.sol";
+import "./CollateralRegistry.sol";
```
Comment (Eenae) [inline contracts/ParametersBatchUpdater.sol:10]: A note: It's preferable to use interfaces.

File: contracts/CollateralRegistry.sol (hunk 1 above)
Comment (Eenae) [reply contracts/CollateralRegistry.sol:52]: Tests should be written covering special cases: deletion of the first and last elements.

**Review by Eenae — CHANGES_REQUESTED** [review]

File: contracts/CollateralRegistry.sol (hunk 1 above)
Comment (bcngod) [reply contracts/CollateralRegistry.sol:52]: Added test cases, works just fine with this logic

Comment (Eenae) [reply contracts/CollateralRegistry.sol:52]: `expect(await this.collateralRegistry.collateralId(collateralAddress3)).to.be.bignumber.equal(new BN(0));`
fails after the removal

File: contracts/ParametersBatchUpdater.sol [hunk 3]
```diff
@@ -5,35 +5,124 @@
 */
 pragma solidity ^0.7.1;
 
-import "./vault-managers/VaultManagerParameters.sol";
-import "./Vault.sol";
-import "./CollateralRegistry.sol";
-import "./oracles/BearingAssetOracleSimple.sol";
+
+interface IVaultManagerParameters {
```
Comment (Eenae) [inline contracts/ParametersBatchUpdater.sol:9]: It makes sense to factor them out as standalone files in `contracts/interfaces` subfolder and reuse (copy-pasted interfaces are still copy-paste).
Also it's better not to omit parameter names.

File: contracts/ParametersBatchUpdater.sol [hunk 4]
```diff
@@ -5,35 +5,124 @@
...
+        require(vaultParameters.canModifyVault(msg.sender), "Unit Protocol: AUTH_FAILED");
+        _;
+    }
+
+    // ensures tx's sender is the Vault
+    modifier onlyVault() {
+        require(msg.sender == vaultParameters.vault(), "Unit Protocol: AUTH_FAILED");
+        _;
+    }
+}
```
Comment (Eenae) [inline contracts/ParametersBatchUpdater.sol:100]: Copy-paste is bad

File: contracts/interfaces/IVault.sol [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+    function borrow ( address asset, address user, uint256 amount ) external returns ( uint256 );
+    function calculateFee ( address asset, address user, uint256 amount ) external view returns ( uint256 );
+    function changeOracleType ( address asset, address user, uint256 newOracleType ) external;
+    function chargeFee ( address asset, address user, uint256 amount ) external;
+    function col (  ) external view returns ( address );
+    function colToken ( address, address ) external view returns ( uint256 );
+    function collaterals ( address, address ) external view returns ( uint256 );
+    function debts ( address, address ) external view returns ( uint256 );
+    function depositCol ( address asset, address user, uint256 amount ) external;
+    function depositEth ( address user ) external;
```
Comment (Eenae) [inline contracts/interfaces/IVault.sol:13]: `payable`

File: contracts/CollateralRegistry.sol [hunk 6]
```diff
@@ -59,6 +59,10 @@ contract CollateralRegistry is Auth {
         emit CollateralRemoved(asset);
     }
 
+    function isCollateral(address asset) external view returns(bool) {
+        return collateralId[asset] != 0 || _collaterals[0] == asset;
```
Comment (Eenae) [inline contracts/CollateralRegistry.sol:63]: Reverts on empty array, but should return `false`:
```solidity
		it('Should add collateral', async function () {
			expect(await this.collateralRegistry.isCollateral(collateralAddress2)).to.equal(false);
```

**Review by Eenae — CHANGES_REQUESTED** [review]

File: contracts/CollateralRegistry.sol [hunk 7]
```diff
@@ -0,0 +1,69 @@
...
+    mapping(address => uint) public collateralId;
+
+    address[] _collaterals;
+    
+    constructor(address _vaultParameters) Auth(_vaultParameters) {}
+
+    function addCollateral(address asset) public onlyManager {
+        require(asset != address(0), "Unit Protocol: ZERO_ADDRESS");
+
+        if (_collaterals.length != 0) {
```
Comment (Eenae) [inline contracts/CollateralRegistry.sol:30]: Btw, here and removeCollateral could be simplified using `isCollateral`. Code reuse is good.

File: contracts/CollateralRegistry.sol [hunk 8]
```diff
@@ -25,7 +25,7 @@ contract CollateralRegistry is Auth {
     constructor(address _vaultParameters) Auth(_vaultParameters) {}
 
     function addCollateral(address asset) public onlyManager {
-        require(asset != address(0), "Unit Protocol: ZERO_ADDRESS");
+        require(!isCollateral(asset), "Unit Protocol: ALREADY_EXISTS");
```
Comment (Eenae) [inline contracts/CollateralRegistry.sol:28]: No-no, not here.
Lines 30-33 and 45.

**Review by Eenae — CHANGES_REQUESTED** [review]

File: contracts/CollateralRegistry.sol [hunk 9]
```diff
@@ -22,7 +22,13 @@ contract CollateralRegistry is Auth {
 
     address[] _collaterals;
     
-    constructor(address _vaultParameters) Auth(_vaultParameters) {}
+    constructor(address _vaultParameters, address[] memory assets) Auth(_vaultParameters) {
+        for (uint i = 0; i < assets.length; i++) {
+            _collaterals.push(assets[i]);
+            collateralId[assets[i]] = i;
```
Comment (Eenae) [inline contracts/CollateralRegistry.sol:28]: Strictly speaking, a duplicate asset check should be added to ensure consistence.

**Review by Eenae — APPROVED** [review]

**Pull request merged**